### DISCUSSION TITLE
How do I access response Body and header inside a function

### DISCUSSION THREAD

**Discussion by ifnotnithya** [thunderclient/thunder-client-support#1281] [Q&A] [answered]
I have a requirement where, 

- I parse the response Body and set multiple values into multiple env variables to be used in subsequent requests. I dont want to manually add these from UI and also there is a possible response variations.
- Parse specific headers and extract values into a variable. 

I am trying to do this in a function. I am not able to figure out how to get the responseBody/headers in the function. 
Appreciate help in this. 

Thanks

**Comment by rangav** [comment, marked as answer]
Hi @ifnotnithya 

please see all the functions and properties available in `tc` object
https://github.com/rangav/thunder-client-support/blob/master/docs/tc-types.d.ts

custom scripting docs
https://github.com/rangav/thunder-client-support/blob/master/docs/filters.md#custom-filters

example script file
https://github.com/rangav/thunder-client-support/blob/master/docs/custom-filters.js

**Comment by ifnotnithya** [comment]
Thanks @rangav 
tc.response is what I was missing. 
Enhancing the functions in my custom filters and think most(if not all) my current use cases should be solved.

**Reply by rangav** [reply]
FYI - If you don't know the launch of our subscriptions - https://www.thunderclient.com/pricing#compare

**Reply by erwinmoretz** [reply]
Hi, 

I have not figured out how to use the tc.response to store the response body inside a variable in a custom filter. Is it possible that you could give me an example of this? :-)

**Reply by rangav** [reply]
```js
function preFilter1() {

    // ---- save json
   var jsonBody = JSON.stringify(tc.response.json);
    tc.setVar("json_body", jsonBody);

// ---- save other response types
   var textBody = tc.response.text;
    tc.setVar("text_body", textBody);
}

module.exports = [preFilter1]
```

Use this vscode will provide autocomplete
```
/// copy tc-types.d.ts file for vscode autocompletion on tc object
/// <reference path="./tc-types.d.ts" />
```

**Reply by rangav** [reply]
you can also use - `// @ts-check` at the top script file 
https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html

**Reply by erwinmoretz** [reply]
thank you for your respond! 

now i figured out, how i can use it. 

for me it was: 

var responseBody = tc.response.json
